src/runtime/update-component.ts

```diff
@@ -48,27 +71,61 @@ const dispatchHooks = (hostRef: d.HostRef, isInitialLoad: boolean) => {
     }
     emitLifecycleEvent(elm, 'componentWillLoad');
     if (BUILD.cmpWillLoad) {
-      promise = safeCall(instance, 'componentWillLoad');
+      // If `componentWillLoad` returns a `Promise` then we want to wait on
+      // whatever's going on in that `Promise` before we launch into
+      // rendering the component, doing other lifecycle stuff, etc. So
+      // in that case we assign the returned promise to the variable we
+      // declared above to hold a possible 'queueing' Promise
+      maybePromise = safeCall(instance, 'componentWillLoad');
     }
   } else {
     emitLifecycleEvent(elm, 'componentWillUpdate');
 
     if (BUILD.cmpWillUpdate) {
-      promise = safeCall(instance, 'componentWillUpdate');
+      // Like `componentWillLoad` above, we allow Stencil component
+      // authors to return a `Promise` from this lifecycle callback, and
+      // we specify that our runtime will wait for that `Promise` to
+      // resolve before the component re-renders. So if the method
+      // returns a `Promise` we need to keep it around!
+      maybePromise = safeCall(instance, 'componentWillUpdate');
     }
   }
 
   emitLifecycleEvent(elm, 'componentWillRender');
   if (BUILD.cmpWillRender) {
-    promise = then(promise, () => safeCall(instance, 'componentWillRender'));
+    maybePromise = enqueue(maybePromise, () => safeCall(instance, 'componentWillRender'));
   }
 
   endSchedule();
-  return then(promise, () => updateComponent(hostRef, instance, isInitialLoad));
+
+  return enqueue(maybePromise, () => updateComponent(hostRef, instance, isInitialLoad));
 };
 
+/**
+ * This function uses a Promise to implement a simple first-in, first-out queue
+ * of functions to be called.
+ *
+ * The queue is ordered on the basis of the first argument. If it's
+ * `undefined`, then nothing is on the queue yet, so the provided funcion can be
+ * called synchronously (although note that this function may return a `Promise`).
+ * The idea is that then the return value of that enqueueing operation is kept
+ * around, so that if it was a `Promise` then subsequent functions can be
+ * enqueued by calling this function again with that `Promise` as the first
+ * argument.
+ *
+ * @param maybePromise either a `Promise` which should resolve before the next function is called or an 'empty' sentinel
+ * @param fn a function to enqueue
+ * @returns either a `Promise` or the return value of the provided function
+ */
+function enqueue(maybePromise: Promise<void> | undefined, fn: () => Promise<void>): Promise<void> | undefined {
```

Comment:
Could we make this an arrow function similar to the original `then` impl please?

Reply:
Sure, we can do arrow + ternary for more compactness